server/asset/dcr/utxo.go

```diff
@@ -65,25 +65,25 @@ func (txio *TXIO) confirmations(ctx context.Context, checkApproval bool) (int64,
 				if err != nil {
 					return -1, err
 				}
-				txio.height = blk.height
-				txio.blockHash = blk.hash
+				txio.height = blk.Height
+				txio.blockHash = *blk.Hash
 			}
 			return verboseTx.Confirmations, nil
 		}
 	} else {
 		// The tx was included in a block, but make sure that the tx's block has
 		// not been orphaned or voted as invalid.
-		mainchainBlock, found := txio.dcr.blockCache.atHeight(txio.height)
-		if !found || mainchainBlock.hash != txio.blockHash {
+		mainchainBlock, found := txio.dcr.blockCache.BlockAt(txio.height)
+		if !found || *mainchainBlock.Hash != txio.blockHash {
 			return -1, ErrReorgDetected
 		}
 		if mainchainBlock != nil && checkApproval {
 			nextBlock, err := txio.dcr.getMainchainDcrBlock(ctx, txio.height+1)
 			if err != nil {
 				return -1, fmt.Errorf("error retrieving approving block tx %s: %w", txio.tx.hash, err)
 			}
-			if nextBlock != nil && !nextBlock.vote {
-				return -1, fmt.Errorf("transaction %s block %s has been voted as invalid", txio.tx.hash, nextBlock.hash)
+			if nextBlock != nil && !nextBlock.Vote {
+				return -1, fmt.Errorf("transaction %s block %s has been voted as invalid", txio.tx.hash, nextBlock.Hash)
```

Comment:
Hadn't considered this recently, but what is the right thing to do if a txn was in a stake-invalidated block?  I think it would be to reset `txio.height` to 0 and try again (making `confirmations` do getrawtransaction) to see if it is now back in mempool or simply in a different block with fewer confirmations.
Right now the various `Confirmations` methods only do the above when the error is `ErrReorgDetected`, but it seems like the proper resolution for stake invalidation too.  Well, actually they makes a new `TXIO` instead of resetting the `height` field, but you know what I mean.
BTW, did you gain any insight about how the infinite recursion happened in that one issue?